File: algorithms/quick-sort/c/quicksort.c
#include <stdio.h>

void swap(int *a, int *b)
{
  int t = *a;
  *a = *b;
  *b = t;
}

int partition(int *a, int l, int h)
{
  int p = a[h];
  int i = l - 1;

  for (int j = l; j < h; j++) {
    if (a[j] < p) {
      i += 1;
      swap(&a[i], &a[j]);
    }
  }

  swap(&a[i + 1], &a[h]);
  return i + 1;
}

void _quicksort(int *a, int l, int h)
{
  if (l < h) {
    int p = partition(a, l, h);
    _quicksort(a, l, p - 1);
    _quicksort(a, p + 1, h);
  }
}

void quicksort(int *a, int n)
{
  return _quicksort(a, 0, n - 1);
}

int main(void)
{
  return 0;
}

File: algorithms/quick-sort/python/quicksort.py
#!/usr/bin/env python3

def _partition(a, lo, hi):
  p = a[hi]
  i = lo - 1

  for j in range(lo, hi):
    if a[j] < p:
      i += 1
      a[i], a[j] = a[j], a[i]

  a[i + 1], a[hi] = a[hi], a[i + 1]
  return i + 1

def _quicksort(a, lo, hi):
  if lo < hi:
    j = _partition(a, lo, hi)
    _quicksort(a, lo, j - 1)
    _quicksort(a, j + 1, hi)

def quicksort(A):
  return _quicksort(A, 0, len(A) - 1)

